lua/ide/panels/panel.lua

```diff
@@ -125,6 +125,24 @@ Panel.new = function(tab, position, components)
 
         vim.api.nvim_buf_set_name(buf, string.format("component://%s:%d:%d", Component.name, panel_win, self.tab))
 
+        -- "lock" the buf to the window with an autocmd
+        vim.api.nvim_create_autocmd({ 'BufWinEnter' }, {
```

Comment:
Could it be more efficient to do this at the Components buffer level? 

For instance, an autocmd can fire only when you enter the Components buffer. Then you can check current Win from there? 

This may be a bit better then firing for ever "BufWinEnter", since we only care to fire this when youre inside a Component buffer/win. Right?

Reply:
That won't work. The whole idea is to fire this autocmd on buffers _other than_ the components' buffers. If a buffer _other than_  the component buffer is opened into a component's _window_ then it has to be moved. If we only fire on component buffers this would not work.

Reply:
`BufWinEnter` also only fires when a different buffer is set as the active buffer in the window, so it's not like it will fire every time you enter a buffer or anything. Only when a buffer is added to a window.

Reply:
Cant we try the above with "BufHidden". 

For instance, 

We only have one window display a components buffer at any time. 

The user accidentally changes the buffer in that win. 

The component's buffer is about to be hidden, now, reset the components win to the components buffer id. 

id would be nice if we can keep to "lock" to the components buffer so we dont fire on every window move.

Reply:
That seems to work, though it has a small downside. Since it fires _before_ the buffer is hidden, we have to wrap the callback in `vim.schedule` so that we can wait until _after_ its hidden in order to get the buffer ID that was just opened, then switch the new buffer to a different window.

This results in the component visually flashing the opened file before switching back, like so:

![CleanShot 2022-12-09 at 07 50 01](https://user-images.githubusercontent.com/8648891/206706381-741409ce-8d5e-4d7e-aa0b-c76d980acd22.gif)

I personally think the current implementation is better.

Reply:
Okay, can you just paste the alternative implementation here so we have an alternative incase we wind up having to do some reeling back on how many aucmds "nvim-ide" fires.

Reply:
Sure, here's the alternative I tried that resulted in the above recording:

```lua
-- "lock" the buf to the window with an autocmd
vim.api.nvim_create_autocmd({ 'BufHidden' }, {
    callback = function()
        vim.schedule(function ()
            if not vim.api.nvim_win_is_valid(panel_win) then
                return
            end
            local panel_win_buf = vim.api.nvim_win_get_buf(panel_win)
            if panel_win_buf ~= buf then
                vim.api.nvim_win_set_buf(panel_win, buf)
                local last_win = self.workspace.get_win()
                vim.api.nvim_win_set_buf(last_win, panel_win_buf)
                vim.api.nvim_set_current_win(last_win)
            end
        end)
    end,
    buffer = buf,
})

```